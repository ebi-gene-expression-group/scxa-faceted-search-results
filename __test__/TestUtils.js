import React from 'react'
import PropTypes from 'prop-types'

const EpisodeCard = ({title}) =>
  <div>
    <p>{title}</p>
  </div>

EpisodeCard.propTypes = {
  title: PropTypes.string.isRequired
}

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
}

const vindicators = [
  { group:`Vindicator`, value: `supernova`, label: `Supernova`, disabled: false},
  { group:`Vindicator`, value: `noob-noob`, label: `Noob-Noob`, disabled: false},
  { group:`Vindicator`, value: `vance_maximus`, label: `Vance Maximus`, disabled: false},
  { group:`Vindicator`, value: `alan_rails`, label: `Alan Rails`, disabled: false},
  { group:`Vindicator`, value: `crocubot`, label: `Crocubot`, disabled: false},
  { group:`Vindicator`, value: `million_ants`, label: `Million Ants`, disabled: false},
  { group:`Vindicator`, value: `morty_smith`, label: `Morty Smith`, disabled: false},
  { group:`Vindicator`, value: `rick_sanchez`, label: `Rick Sanchez`, disabled: false},
  { group:`Vindicator`, value: `lady_katana`, label: `Lady Katana`, disabled: false},
  { group:`Vindicator`, value: `calypso`, label: `Calypso`, disabled: false},
  { group:`Vindicator`, value: `diablo_verde`, label: `Diablo Verde`, disabled: false}
]

const episodes = [
  {
    element: {
      title: `Raising Gazorpazorp`
    },
    facets: [
      {
        group: `Planet`,
        value: `gazorpazorp`,
        label: `Gazorpazorp`,
        description: `Planet Tooltip`
      },
      {
        group: `Guest character`,
        value: `gwendolyn`,
        label: `Gwendolyn`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `ma-sha`,
        label: `Ma-Sha`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `The wedding squanchers`
    },
    facets: [
      {
        group: `Planet`,
        value: `squanch`,
        label: `Squanch`,
        description: `Planet Tooltip`
      },
      {
        group: `Guest character`,
        value: `birdperson`,
        label: `Birdperson`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `squanchy`,
        label: `Squanchy`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `2`,
        label: `Season 2`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `The Rickshank redemption`
    },
    facets: [
      {
        group: `Guest character`,
        value: `birdperson`,
        label: `Birdperson`,
        description: `Guest character Tooltip`
      },
      {
        group: `Planet`,
        value: `buttworld`,
        label: `Buttworld`,
        description: `Planet Tooltip`
      },
      {
        group: `Season`,
        value: `3`,
        label: `Season 3`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `Ricksy business`
    },
    facets: [
      {
        group: `Guest character`,
        value: `squanchy`,
        label: `Squanchy`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `abradolf_lincler`,
        label: `Abradolf Lincler`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `Close Rick-counters of the Rick kind`
    },
    facets: [
      {
        group: `Guest character`,
        value: `ricktiminus_sancheziminius`,
        label: `Ricktiminus Sancheziminius`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `abradolf_lincler`,
        label: `Abradolf Lincler`,
        description: `Guest character Tooltip`
      },
      {
        group: `Planet`,
        value: `buttworld`,
        label: `Buttworld`,
        description: `Planet Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  }
]

export {getRandomInt, vindicators, episodes, EpisodeCard}
