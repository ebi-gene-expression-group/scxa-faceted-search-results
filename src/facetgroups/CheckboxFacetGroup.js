import React from 'react'
import { xorBy as _xorBy } from 'lodash'

import FacetGroupPropTypes from './FacetGroupPropTypes'

const CheckboxOption = ({group, value, label, disabled, onChange, checked}) =>
  <div>
    <input type={`checkbox`} {...{value, checked, disabled}} onChange={() => onChange({group, label, value})}/>
    <label style={disabled ? {color: `lightgrey`} : {}}>{label}</label>
  </div>

// In principle we donâ€™t need this component to be stateful, but in doing so and returning a (facetGroupName, facets)
// we have the same API as React-Select and we can have the same callback for both checkbox-style facet groups and
// multiselect dropdowns
class CheckboxFacetGroup extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      checkedFacets: []
    }
    this._handleChange = this._handleChange.bind(this)
  }

  _handleChange(facet) {
    this.setState(
      { checkedFacets: _xorBy(this.state.checkedFacets, [facet], `value`) },
      () => this.props.onChange(facet.group, this.state.checkedFacets))
  }

  render() {
    const {facetGroupName, facets} = this.props
    const {checkedFacets} = this.state

    return (
      <div className={`padding-bottom-xlarge`}>
        <h4>{facetGroupName}</h4>
        {facets.map((facet) =>
          <CheckboxOption {...facet}
                          onChange={this._handleChange}
                          key={facet.value}
                          checked={checkedFacets.some((checkedFacet) => checkedFacet.value === facet.value)}/>
        )}
      </div>
    )
  }
}

CheckboxFacetGroup.propTypes = FacetGroupPropTypes

export default CheckboxFacetGroup
