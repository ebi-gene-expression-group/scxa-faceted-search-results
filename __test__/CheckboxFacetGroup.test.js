import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { getRandomInt, episodes } from './TestUtils'

import CheckboxFacetGroup from '../src/facetgroups/CheckboxFacetGroup'

Enzyme.configure({ adapter: new Adapter() })

describe(`CheckboxFacetGroup`, () => {
  const allFacets = episodes.reduce((acc, episode) => acc.concat(episode.facets), [])
  const uniqueFacets =
    allFacets
      .filter((facet, index) => allFacets.findIndex((thatFacet) => facet.value === thatFacet.value) === index)
      .map((facet) => ({
        ...facet,
        disabled: false
      }))

  const props = {
    facetGroupName: `Planet`,
    facetGroupNameDescription: `Planets where the events of this episode take place`,
    facets: uniqueFacets,
    onChange: () => {}
  }

  const noTooltipProps = {
    facetGroupName: `Planet`,
    facets: [{
      group: `Guest character`,
      value: `birdperson`,
      label: `Birdperson`
    }],
    onChange: () => {}
  }

  test(`checks whether tooltip exists`, () => {
    const groups = [...new Set(uniqueFacets.map((facet) => facet.group))]
    const descriptions = [...new Set(uniqueFacets.map((facet) => facet.description))]
    const randomCheckboxFacetGroup = groups[getRandomInt(0, groups.length)]
    const randomCheckboxFacetGroupDescription = descriptions[getRandomInt(0, descriptions.length)]

    const wrapper =
      mount(
        <CheckboxFacetGroup
          facetGroupName= {randomCheckboxFacetGroup}
          facetGroupNameDescription={randomCheckboxFacetGroupDescription}
          facets={props.facets}
          onChange={props.onChange}
          key={randomCheckboxFacetGroup} />)
    expect(wrapper.find(`sup`).first().html()).toEqual(expect.stringMatching(`data-tooltip`))
  })

  test(`checks tooltip does not exist when no tooltip text in payload`, () => {
    const wrapper =
      mount(
        <CheckboxFacetGroup
          facetGroupName= {noTooltipProps.facetGroupName}
          facetGroupNameDescription={null}
          facets={noTooltipProps.facets}
          onChange={noTooltipProps.onChange}
          key={noTooltipProps.facetGroupName} />)
    expect(wrapper.find(`sup`).exists()).toEqual(false)
  })

  test(`matches snapshot`, () => {
    const tree =
      renderer.create(
        <CheckboxFacetGroup
          facetGroupName= {props.facetGroupName}
          facetGroupNameDescription={props.facetGroupNameDescription}
          facets={props.facets}
          onChange={props.onChange}
          key={props.facetGroupName} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
