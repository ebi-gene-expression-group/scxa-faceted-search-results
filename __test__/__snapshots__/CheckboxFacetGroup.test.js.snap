// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFacetGroup matches snapshot 1`] = `
<div
  className="padding-bottom-xlarge"
>
  <h4>
    Planet
    <span>
      <sup
        aria-haspopup="true"
        className="has-tip tip-right"
        data-tooltip={true}
        style={
          Object {
            "background": "white",
            "border": "none",
          }
        }
        title="Planet tooltip"
      >
        ?
      </sup>
    </span>
  </h4>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="gazorpazorp"
    />
    <label
      style={Object {}}
    >
      Gazorpazorp
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="gwendolyn"
    />
    <label
      style={Object {}}
    >
      Gwendolyn
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="ma-sha"
    />
    <label
      style={Object {}}
    >
      Ma-Sha
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="1"
    />
    <label
      style={Object {}}
    >
      Season 1
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="squanch"
    />
    <label
      style={Object {}}
    >
      Squanch
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="birdperson"
    />
    <label
      style={Object {}}
    >
      Birdperson
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="squanchy"
    />
    <label
      style={Object {}}
    >
      Squanchy
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="2"
    />
    <label
      style={Object {}}
    >
      Season 2
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="buttworld"
    />
    <label
      style={Object {}}
    >
      Buttworld
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="3"
    />
    <label
      style={Object {}}
    >
      Season 3
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="abradolf_lincler"
    />
    <label
      style={Object {}}
    >
      Abradolf Lincler
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="ricktiminus_sancheziminius"
    />
    <label
      style={Object {}}
    >
      Ricktiminus Sancheziminius
    </label>
  </div>
</div>
`;
